src/StoreApi/Utilities/CartController.php

```diff
@@ -355,15 +355,52 @@ private function get_error_message_for_stock_exception_type( $exception_type, $s
 		return __( 'There was an error with an item in your cart.', 'woo-gutenberg-products-block' );
 	}
 
+	/**
+	 * Validate cart and check for errors.
+	 *
+	 * @throws InvalidCartException Exception if invalid data is detected in the cart.
+	 * @throws RouteException Exception if invalid data is detected in the cart.
+	 */
+	public function validate_cart() {
+
+		$cart = $this->get_cart_instance();
+		$this->validate_cart_items();
+		$this->validate_cart_coupons();
+
+		/**
+		 * Fire action to validate cart. Functions hooking into this should throw a \InvalidCartException.
+		 *
+		 * @example See docs/examples/validate-cart.md
+		 *
+		 * @param \WC_Cart $cart Cart object.
+		 */
+		do_action( '__experimental_woocommerce_store_api_validate_cart', $cart );
```

Comment:
I can see why you've done this, but this does force a 3PD to have knowledge of both `WP_Error` handling, as well as `InvalidCartException`. 

At the very least, we could provide the `InvalidCartException` class/namespace via this hook so that a 3PD can throw the error without needing to know the correct namespace. This also guards against future updates if a namespace were to change.

The alternative that we should perhaps consider would be to rather than force 3PD to throw the exception with a `WP_Error` object, make this a filter hook which passes through a `WP_Error` object. 3PD could then use `$error->add()` to add errors to the shared instance.

This method (`validate_cart`) could also be further enhanced to replace `convert_notices_to_exceptions` to a new `convert_notices_to_errors` method which again, appends more errors to the shared `WP_Error` object.

Finally, once all this has completed, the generic `InvalidCartException` can be thrown internally.

The advantage of this approach would be:

1. 3PD only need knowledge of `WP_Error`, not the internals of our exception handling
2. 3PD can add/return multiple error messages
3. All messages + notices would be converted to a single Exception with the full list of errors which will be more useful for the response.

We also don't block the usage of Exceptions should there be a critical issue that would require no other logic or errors to be thrown/added. We leave that down to the 3PD.

Thoughts?

Reply:
It took me a while to wrap my head around this. 

> [...] this force a 3PD to have knowledge of both WP_Error handling, as well as InvalidCartException.

Excellent point; having to handle both seems a bit overwhelming. Also, the namespace-"parade" is a commitment. 

> make this a filter hook which passes through a WP_Error object. 3PD could then use $error->add() to add errors to the shared instance.

Suppose we go in this direction, which seems like a promising alternative and, I'd say, much cleaner; we need to re-think the whole validation sequence.

**Introducing a shared wp error instance**

Keeping all the exceptions internally sounds reasonable if we _green-light_ the following:
- The `validate_xxx` methods will be responsible for _throwing_ all the exceptions.
- The `validate_xxx` methods should be clean from _try...catch_ blocks to further highlight this architecture.
- Exceptions will be caught from (a) the error getters and (b) the general `AbstractCartRoute::get_response` method as our last resort.

From a 3PD perspective, I would say that dealing with a filter can sometimes lead to unwanted effects:
- If you return nothing/null, the system will need to handle this type of validation.
- If you return a new `WP_Error` instance, that will "replace" all previous messages.

These seem acceptable as long as we only group all the 3PDs messages and the legacy notices.

As far as I understand, the broad idea behind the StoreAPI system is that someone can throw a `RouteException` that will end up in some errors array or UI context at any moment of new or existing legacy code. 

Definitely agree that we need to _draw the line_ on where and which exceptions should be publicly considered best practice.

This approach can simplify things for 3PDs by leaving them the following options:
- Throw a `RouteException` if things are going bad at anywhere, or;
- Return a `WP_Error` instance for multiple errors handling in validation-specific filters' context.

These options explain that we need to decide which points/hooks will be "_packing_" the "multi-errors" feature versus the single `RouteException` error.

We could introduce the shared `WP_Error` handling for more hooks than the `__experimental_woocommerce_store_api_validate_cart` to keep things consistent. For example:
- `wooocommerce_store_api_validate_cart_item`
- maybe more?

Moreover, the system will use the legacy(?) `woocommerce_coupon_is_valid` to validate coupons, so in order to be aligned, it should make sense to introduce a new `wooocommerce_store_api_validate_cart_coupon`. This could also leverage the shared wp error approach.

Thinking ahead probably, so I'll sum up:
- I agree that exposing more exceptions than the `RouteException` seems too _technical_ for 3PDs; the shared wp error keeps the technical barrier a bit lower, which is always a good thing.
- Handling multi-errors in specific only contexts using `WP_Error` instances sounds enjoyable.
- Having the ability to streamline error handling by throwing any known exception is a plus.
- If we go there, we probably need to commit to this approach in all related places in the code. 

Thoughts? 

_PS: I will be merging with `trunk` soon and refactoring the versioning namespace, but I got stuck on some unrelated PHPCS errors after the merge -- will look into this tomorrow._

Reply:
> I'd probably stick to one action or filter hook for the 3PD aspect of this - surfacing a WP_Error Object which they can append to.

One filter sounds like a good plan. We are also introducing this new error-handling wrapped in an 'experimental__x' filter, making even more sense.

> We cannot get around the usage of legacy hooks, so in those instances I'd just work with whatever data type they return and convert to exceptions or WP_Error

Sounds good! Given that we can now handle multiple errors, a part of integration work for several plugins could be _plug-n-play_ with this approach.

> The alternative which might work better is a registration system where you can register a callback. 

As far as I understand, you are suggesting a custom `apply_filters` mechanism? Only this would be strict-typed, or is there perhaps an easier way to accomplish this?

I like this approach!

Dealing with recursive functions _(filters-in-filters, etc.)_ could require creating a new Utility class that will extend the `WP_Hook` class and introduce a method e.g., `apply_filters_to_wp_error()`. 

This pathway:
- Offers a simple way for 3PDs to use the `add_filter()` function.
- Protects the system and ensures that everybody will get a `WP_Error` argument no matter what they return.
- Covers recursive behavior and keeps the global `$wp_filters` intact.
- Has low maintenance due to this class's low frequency of changes -- but not _zero_, unfortunately.
- It leaves the ability for 3PDs to throw a `RouteException` to avoid the error-handling group.

I hope I'm not missing an obvious alternative here.

> It leaves the ability for 3PDs to throw a `RouteException` to avoid the error-handling group.

If we start gathering all errors from third parties, we may end up with a _not aesthetic_ number of error notices in the UI.

However, someone could throw a `RouteException` in the validate-filter context; this is a feature/bug that can be used to bypass the error grouping UX issue and respond with a _single_ error string. 

Trying to group and display all 3PDs errors/notices, **without the one interrupting the other** seems like a solid trade-off for this possible UX issue, and a nice plan to promote. Having the exception _shortcut_ is also always a plus.

Reply:
We can probably leave the error handling to the client. As long as we can distinguish errors by their error codes. Critical errors should probably just throw Exceptions however, so there is only one shown.

> As far as I understand, you are suggesting a custom apply_filters mechanism? Only this would be strict-typed, or is there perhaps an easier way to accomplish this?

To clarify what I meant, you'd have a single `apply_filter` named something like `cart_validation_callbacks`, and the consumers would return functions there. We'd have an array of callbacks to process.

Thinking about it more though, that is still not ideal because then 3PD could destroy each others callbacks, so if we do something like this we might want a `registration` function instead (no filters), maybe similar, or part of, the ExtendRestAPI class approach: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/Domain/Services/ExtendRestApi.php#L291

Overcomplicated?

Reply:
> Overcomplicated?

Yes and no. In my opinion, having the opportunity to start over and reduce the frustration is a big plus for us.

On the plus side, if we can introduce a way that everyone would get a `WP_Error` to work with, without needing to deep dive into the hooks sequence to search for the _one_ piece of code that ruins everything, I'd say let's go.

On the other hand, all this is starting to feel like we are moving forward from the typical `apply_filters` system, which isn't necessarily bad. Some thoughts on this approach, though:
- 3PDs can no longer use a _priority_ in their callbacks; this could lead to a race to who comes first.
- Requires a new way of _hooking-in_; instead of traditional filters and actions.
- Could lead to a technical barrier that is equally frustrating as we started this analysis (having to deal with exceptions).

We have to decide on this trade-off here. We could also develop a global Utility class for handling "callback queues" and reuse it. But the main question is about _convenience_ and _robustness_ of extensibility needs.

In my opinion, this custom payments callback queue is an excellent piece of work when handling complex data/schema operations; I'm just not sure if we need this in a more straightforward case like this. Also, the need for a validation process may be more complex when dealing with multiple business rules. 

For example, someone might need to wait for "Points and Rewards"-discounts to kick in before validating his own thing. This example should highlight more the need for having the _priority_ feature.

Thoughts? A system that will inherit the `WP_Hook` class could work here, but I definitely understand that this might also be "too complicated." 🤔

Reply:
To sum up our options here:
- Introduce a validation callback queue and perhaps revisit later to introduce _priority_ or even a Utility _singleton_.
- Work with traditional filters and leave the type handling to everyone's need to _follow the rules_.

At this point, both should work with or without revisits, but the first one is a bit _radical_.

Reply:
@xristos3490 I toyed with a callback registry for this (fire action passing registry class, others use `$registry->register()` to add a callback, then callbacks are processed) and it worked but I think it is a bit convoluted...

I think all things considered (trying to balance robustness with each of use), let's do a **filter** hook with the following caveats:

1. The filter hook provides a `WP_Error` object. Extensions can add errors to this object for later handling.
2. The filter hook is wrapped in try/catch to catch any exceptions.
3. The filter hook must return `WP_Error` - if it doesn't we need an exception.

Along these lines:

```php
$cart_errors = apply_filters( '__experimental_woocommerce_store_api_cart_errors', new \WP_Error() );

if ( ! is_wp_error( $cart_errors ) ) {
	// Throw exception.
}

if ( $cart_errors->has_errors() ) {
	throw new RouteException( 'cart_error', $cart_errors->get_error_message() );
}
```

We'll need handling for multiple errors (I think we already have work in place for this), and it would be consumed like:

```php
add_action(
				'__experimental_woocommerce_store_api_cart_errors',
				function( \WP_Error $errors ) {
					$errors->add( 'some_error', 'I am broken.' );
					return $errors;
				}
			);
```

Reply:
I also experimented with a simple queue runner logic, and I concluded to the same judgment. Here's the main code part for reference:
```php
foreach ( $queue as $cb ) {
    if ( ! is_callable( $cb ) ) {
        continue;
    }

    $error = call_user_func_array( $cb, array( $cart ) );
    if ( is_a( $error, 'WP_Error' ) ) {
        // Restore.
        $global_error->merge_from( $error );
    }
}
```
One could potentially find their way into "hacking" [this](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5904/commits/792623392d9c06bd32049fd96580c812b0ec4779) system with the same severity as the legacy hooks system. So, I definitely agree with your spottings! 

I did a [commit](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5904/commits/501132bdde0e3624f5e36d136c1e5192ab9201f0) that adds the `__experimental_woocommerce_store_api_cart_errors
` filter.

> The filter hook is wrapped in try/catch to catch any exceptions.

Can you please elaborate on this? I think that we could leave the filter unwrapped. This way, the corresponding wrapper _(either the cart getter or the route response handler)_ will catch all the exceptions.

Other notes:
- Notices are grouped in a single `WP_Error` instance and handled accordingly. @ commit
- The previous `NoticeHandler::convert_notices_to_exceptions` is removed -- perhaps we need to add a deprecation warning?
- We could deprecate the previous error getters -- see this [comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5904/files#r816006972)

Lastly, what are your thoughts about the existing `wooocommerce_store_api_validate_cart_item` hook?

Every purpose of this action could probably be handled using the new  `__experimental_woocommerce_store_api_cart_errors`, and we could safely remove/deprecate it to avoid confusion.

At this point, the code in this PR should be aligned with our discussion.